In [42]:
import pandas as pd
from sqlalchemy import create_engine
import os
import numpy as np


In [43]:
csv_file = os.path.join( "Resources", "president_polls.csv")
polls_data_df = pd.read_csv(csv_file)
polls_data_df.head()

question_id  poll_id  cycle           state  pollster_id  \
0       134285    71678   2020    Pennsylvania          383   
1       134285    71678   2020    Pennsylvania          383   
2       134286    71679   2020        Michigan          383   
3       134286    71679   2020        Michigan          383   
4       134249    71662   2020  North Carolina          399   

                                   pollster sponsor_ids  \
0                     Public Policy Polling        1484   
1                     Public Policy Polling        1484   
2                     Public Policy Polling        1484   
3                     Public Policy Polling        1484   
4  Rasmussen Reports/Pulse Opinion Research         NaN   

                       sponsors                              display_name  \
0  American Bridge 21st Century                     Public Policy Polling   
1  American Bridge 21st Century                     Public Policy Polling   
2  American Bridge 21st Century                     Public Policy Polling   
3  American Bridge 21st Century                     Public Policy Polling   
4                           NaN  Rasmussen Reports/Pulse Opinion Research   

   pollster_rating_id  ...        created_at notes  \
0               263.0  ...  10/22/2020 15:33   NaN   
1               263.0  ...  10/22/2020 15:33   NaN   
2               263.0  ...  10/22/2020 15:33   NaN   
3               263.0  ...  10/22/2020 15:33   NaN   
4               277.0  ...  10/22/2020 10:27   NaN   

                                                 url    stage race_id answer  \
0  https://americanbridgepac.org/wp-content/uploa...  general    6249  Biden   
1  https://americanbridgepac.org/wp-content/uploa...  general    6249  Trump   
2  https://americanbridgepac.org/wp-content/uploa...  general    6233  Biden   
3  https://americanbridgepac.org/wp-content/uploa...  general    6233  Trump   
4  https://www.rasmussenreports.com/public_conten...  general    6238  Biden   

  candidate_id       candidate_name  candidate_party   pct  
0        13256  Joseph R. Biden Jr.              DEM  51.0  
1        13254         Donald Trump              REP  46.0  
2        13256  Joseph R. Biden Jr.              DEM  50.0  
3        13254         Donald Trump              REP  43.0  
4        13256  Joseph R. Biden Jr.              DEM  48.0  

[5 rows x 38 columns]

In [57]:
new_polls_data_df = polls_data_df[[ 'state', 'pollster_id', 'pollster', 'display_name', 'fte_grade', 'sample_size', 'population', 'methodology', 'end_date', 'nationwide_batch', 'answer', 'candidate_id', 'candidate_party', 'pct' ]].copy()
new_polls_data_df.head()

state  pollster_id                                  pollster  \
0    Pennsylvania          383                     Public Policy Polling   
1    Pennsylvania          383                     Public Policy Polling   
2        Michigan          383                     Public Policy Polling   
3        Michigan          383                     Public Policy Polling   
4  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

                               display_name fte_grade  sample_size population  \
0                     Public Policy Polling         B        890.0          v   
1                     Public Policy Polling         B        890.0          v   
2                     Public Policy Polling         B        804.0          v   
3                     Public Policy Polling         B        804.0          v   
4  Rasmussen Reports/Pulse Opinion Research        C+        800.0         lv   

  methodology    end_date  nationwide_batch answer  candidate_id  \
0    IVR/Text  10/22/2020             False  Biden         13256   
1    IVR/Text  10/22/2020             False  Trump         13254   
2    IVR/Text  10/22/2020             False  Biden         13256   
3    IVR/Text  10/22/2020             False  Trump         13254   
4  IVR/Online  10/21/2020             False  Biden         13256   

  candidate_party   pct  
0             DEM  51.0  
1             REP  46.0  
2             DEM  50.0  
3             REP  43.0  
4             DEM  48.0

In [47]:
# new_polls_data_df['row_id'] = np.arange(len(new_polls_data_df))
# new_polls_data_df.head()


state  pollster_id                                  pollster  \
0    Pennsylvania          383                     Public Policy Polling   
1    Pennsylvania          383                     Public Policy Polling   
2        Michigan          383                     Public Policy Polling   
3        Michigan          383                     Public Policy Polling   
4  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

                               display_name fte_grade  sample_size population  \
0                     Public Policy Polling         B        890.0          v   
1                     Public Policy Polling         B        890.0          v   
2                     Public Policy Polling         B        804.0          v   
3                     Public Policy Polling         B        804.0          v   
4  Rasmussen Reports/Pulse Opinion Research        C+        800.0         lv   

  methodology    end_date  nationwide_batch answer  candidate_id  \
0    IVR/Text  10/22/2020             False  Biden         13256   
1    IVR/Text  10/22/2020             False  Trump         13254   
2    IVR/Text  10/22/2020             False  Biden         13256   
3    IVR/Text  10/22/2020             False  Trump         13254   
4  IVR/Online  10/21/2020             False  Biden         13256   

  candidate_party   pct  row_id  
0             DEM  51.0       0  
1             REP  46.0       1  
2             DEM  50.0       2  
3             REP  43.0       3  
4             DEM  48.0       4

# Which colums have null data?

In [58]:
new_polls_data_df.isnull().any()

state                True
pollster_id         False
pollster            False
display_name        False
fte_grade            True
sample_size          True
population          False
methodology          True
end_date            False
nationwide_batch    False
answer              False
candidate_id        False
candidate_party     False
pct                 False
dtype: bool

**Columns with null values: state, fte_grade, sample_size, and methodology**

In [59]:
new_polls_data_df.count()

state                8184
pollster_id         12812
pollster            12812
display_name        12812
fte_grade           11865
sample_size         12810
population          12812
methodology         12783
end_date            12812
nationwide_batch    12812
answer              12812
candidate_id        12812
candidate_party     12812
pct                 12812
dtype: int64

In [60]:
new_polls_data_df.isnull().any(axis=1)

0        False
1        False
2        False
3        False
4        False
         ...  
12807    False
12808    False
12809    False
12810     True
12811     True
Length: 12812, dtype: bool

In [61]:
new_polls_data_df[new_polls_data_df.isnull().any(axis=1)]

state  pollster_id                                  pollster  \
8       NaN         1193                              SurveyMonkey   
9       NaN         1193                              SurveyMonkey   
10      NaN         1193                              SurveyMonkey   
11      NaN         1193                              SurveyMonkey   
12      NaN          228                                  IBD/TIPP   
...     ...          ...                                       ...   
12799   NaN         1416               Harris Insights & Analytics   
12800   NaN         1416               Harris Insights & Analytics   
12801   NaN         1416               Harris Insights & Analytics   
12810   NaN          399  Rasmussen Reports/Pulse Opinion Research   
12811   NaN          399  Rasmussen Reports/Pulse Opinion Research   

                                   display_name fte_grade  sample_size  \
8                                  SurveyMonkey        D-      13943.0   
9                                  SurveyMonkey        D-      13943.0   
10                                 SurveyMonkey        D-      13943.0   
11                                 SurveyMonkey        D-      13943.0   
12                                     IBD/TIPP       A/B        965.0   
...                                         ...       ...          ...   
12799                                   HarrisX         C       1001.0   
12800                                   HarrisX         C       1001.0   
12801                                   HarrisX         C       1001.0   
12810  Rasmussen Reports/Pulse Opinion Research        C+       1000.0   
12811  Rasmussen Reports/Pulse Opinion Research        C+       1000.0   

      population        methodology    end_date  nationwide_batch    answer  \
8             rv             Online  10/21/2020             False     Biden   
9             rv             Online  10/21/2020             False     Trump   
10            lv             Online  10/21/2020             False     Biden   
11            lv             Online  10/21/2020             False     Trump   
12            lv  Live Phone/Online  10/21/2020             False     Biden   
...          ...                ...         ...               ...       ...   
12799         rv             Online  12/17/2018             False     Trump   
12800         rv             Online  12/17/2018             False  O'Rourke   
12801         rv             Online  12/17/2018             False     Trump   
12810         lv         IVR/Online  11/13/2018             False     Obama   
12811         lv         IVR/Online  11/13/2018             False     Trump   

       candidate_id candidate_party   pct  
8             13256             DEM  52.0  
9             13254             REP  45.0  
10            13256             DEM  52.0  
11            13254             REP  46.0  
12            13256             DEM  49.6  
...             ...             ...   ...  
12799         13254             REP  37.0  
12800         13276             DEM  30.0  
12801         13254             REP  37.0  
12810         13253             DEM  50.0  
12811         13254             REP  43.0  

[5215 rows x 14 columns]

**There are 5,215 rows out of 12,812 rows with null data. We note this but will keep null data for now**

# Evaluate what type of data types are there?

In [62]:
new_polls_data_df.dtypes

state                object
pollster_id           int64
pollster             object
display_name         object
fte_grade            object
sample_size         float64
population           object
methodology          object
end_date             object
nationwide_batch       bool
answer               object
candidate_id          int64
candidate_party      object
pct                 float64
dtype: object

In [66]:
output_file = os.path.join("Output", "polls_clean_df.csv")

new_polls_data_df.to_csv(output_file, index=False, header=True)

# Connect to local database

In [67]:
pg_user = 'postgres'
pg_password = 'Rouzbeh237*tx'
db_name = 'POTUS_db'

connection_string = f"{pg_user}:{pg_password}@localhost:5432/{db_name}"
engine = create_engine(f'postgresql://{connection_string}')

In [68]:
engine.table_names()

['polls']

In [69]:
pd.read_sql_query('select * from polls', con=engine).head()

id           state  pollster_id                                  pollster  \
0   1    Pennsylvania          383                     Public Policy Polling   
1   2    Pennsylvania          383                     Public Policy Polling   
2   3        Michigan          383                     Public Policy Polling   
3   4        Michigan          383                     Public Policy Polling   
4   5  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

                               display_name fte_grade  sample_size population  \
0                     Public Policy Polling         B        890.0          v   
1                     Public Policy Polling         B        890.0          v   
2                     Public Policy Polling         B        804.0          v   
3                     Public Policy Polling         B        804.0          v   
4  Rasmussen Reports/Pulse Opinion Research        C+        800.0         lv   

  methodology    end_date  nationwide_batch answer  candidate_id  \
0    IVR/Text  2020-10-22             False  Biden         13256   
1    IVR/Text  2020-10-22             False  Trump         13254   
2    IVR/Text  2020-10-22             False  Biden         13256   
3    IVR/Text  2020-10-22             False  Trump         13254   
4  IVR/Online  2020-10-21             False  Biden         13256   

  candidate_party   pct  
0             DEM  51.0  
1             REP  46.0  
2             DEM  50.0  
3             REP  43.0  
4             DEM  48.0